### PR TITLE
add details, fix spot access

### LINKED ISSUE #7: "*" resources is not a too wide permission?
Here: https://github.com/kubecost/cloudformation/blob/4d5c48ea7009a208a67c727203f8052c342f052d/kubecost-sub-account-permissions.yaml#L71

### PR DESCRIPTION
**Pull request by JesseGoodier**
Fixes #7

### CODE REVIEW

**Comment by jessegoodier** [comment]
@srpomeroy

File: kubecost-single-account-permissions.yaml [hunk 1]
```diff
@@ -109,5 +109,5 @@ Resources:
           Action:
           - organizations:ListAccounts
           - organizations:ListTagsForResource
-          Resource: 
-          - "*"
+          Resource:
+          - !Sub 'arn:aws:s3:::${SpotDataFeedBucketName}*'
```
Comment (srpomeroy) [inline kubecost-single-account-permissions.yaml]: A `*` without a prefacing `/` is a wild card on the name, not the objects.
Though we may need to add a second resource for just the bucket.

```suggestion
          - !Sub 'arn:aws:s3:::${SpotDataFeedBucketName}/*'
```

File: kubecost-single-account-permissions.yaml [hunk 2]
```diff
@@ -5,10 +5,10 @@ Description: Kubecost Masterpayer Account Template
 
 Parameters:
   AthenaCURBucket:
-    Description: The bucket where the CUR is sent from the “Setting up the CUR” step.
+    Description: The bucket where the CUR is sent from the "Setting up the CUR" step. This is in the AWS console>Billing Dashboard>Cost & usage reports>S3 Bucket for the daily report. https://docs.kubecost.com/install-and-configure/install/cloud-integration/aws-cloud-integrations#step-1-setting-up-the-cur
     Type: String
   SpotDataFeedBucketName:
-    Description: Optional. The bucket where the spot data feed is sent from the “Setting up the Spot Data feed” step.
+    Description: Optional- Kubecost uses actual billing for all cloud resources regardless of the spot feed being available. With the spot feed, near-term (48h) reports will use the spot rates. The bucket where the spot data feed is sent from the "Spot Data feed integration" step. This can be done at any time. https://docs.kubecost.com/install-and-configure/install/cloud-integration/aws-cloud-integrations/aws-spot-instances
```
Comment (jessegoodier) [inline kubecost-single-account-permissions.yaml]: All for suggestions on this, it is a bit wordy.

File: kubecost-sub-account-permissions.yaml [hunk 3]
```diff
@@ -68,4 +68,5 @@ Resources:
           - s3:HeadObject
           - s3:List*
           - s3:Get*
-          Resource: "*"
+          Resource:
+          - !Sub 'arn:aws:s3:::${SpotDataFeedBucketName}*'
```
Comment (srpomeroy) [inline kubecost-sub-account-permissions.yaml:72]: A `*` without a prefacing `/` is a wild card on the name, not the objects.
Though we may need to add a second resource for just the bucket.

```suggestion
          - !Sub 'arn:aws:s3:::${SpotDataFeedBucketName}/*'
```

**Review by srpomeroy — COMMENTED** [review]
We should verify the following.

File: kubecost-single-account-permissions.yaml (hunk 1 above)
Comment (jessegoodier) [reply kubecost-single-account-permissions.yaml]: this was the wrong location in this file for spot. Fixed

File: kubecost-single-account-permissions.yaml (hunk 2 above)
Comment (srpomeroy) [reply kubecost-single-account-permissions.yaml]: Why don't we link to docs?
https://docs.kubecost.com/install-and-configure/install/cloud-integration/aws-cloud-integrations/aws-spot-instances

Comment (jessegoodier) [reply kubecost-single-account-permissions.yaml]: it is there, at the end.

Are you thinking- just link to docs without more description?

**Comment by jessegoodier** [comment]
@srpomeroy 
needs some testing given the number of changes and I don't trust myself 👍

**Comment by srpomeroy** [comment]
> @srpomeroy needs some testing given the number of changes and I don't trust myself 👍

Looking good. I can test tomorrow.

**Comment by jessegoodier** [comment]
> @srpomeroy any news?

thanks for the reminder on this one. I'm somewhat surprised that nobody else has reported the issue.

In the short-term, this guide is current: https://docs.kubecost.com/install-and-configure/install/cloud-integration/aws-cloud-integrations/aws-cur-setup

does this work for you?

**Comment by pierluigilenoci** [comment]
hopes-and-dreams-barney-hopes

**Comment by spkane** [comment]
These files are still directly referenced in the latest documentation at https://docs.kubecost.com/install-and-configure/install/cloud-integration/aws-cloud-integrations#step-3-setting-up-iam-permissions

which I am actively trying to setup now, so it would be great to get any required fixes merged if they are needed.

**Comment by spkane** [comment]
@cliffcolvin Any idea who could confirm that these are all still accurate and could also test and merge some fixes?

**Comment by jessegoodier** [comment]
Thanks for the note @spkane, without feedback (a while ago)- I had forgotten people use this.

have you seen this guide?
https://docs.kubecost.com/install-and-configure/install/cloud-integration/aws-cloud-integrations/aws-cloud-integration-using-irsa

It is the most current. It can be used even if Kubecost runs in your payer account, just skip the step for granting cross account privs.

I'll make a task to clean up those docs and references to this repo.

**Comment by spkane** [comment]
> have you seen this guide? https://docs.kubecost.com/install-and-configure/install/cloud-integration/aws-cloud-integrations/aws-cloud-integration-using-irsa
> 
> It is the most current. It can be used even if Kubecost runs in your payer account, just skip the step for granting cross account privs.

I'll take a look over it and try that approach. To be honest, Although I think I have the details mostly worked out, I've found the documentation lacking in detail. It is not necessarily that anything is missing, but just that the language isn't as specific as it could be, so it is a bit hard determine the corrects steps you should be doing based on your setup, which, in our case, is multiple sub-accounts that each has EKS clusters in them.

> I'll make a task to clean up those docs and references to this repo.

Thank you that is appreciated. At a minimum, either archiving, or at least correcting the files in here would be good. I added an issue, about the smart quotes that are in the descriptions, and should be replaced.

**Comment by jessegoodier** [comment]
@spkane we will test these next week.

It sounds like they worked for you, other than the smart quotes?

**Comment by spkane** [comment]
> @spkane we will test these next week.
> 
> It sounds like they worked for you, other than the smart quotes?

@jessegoodier Honestly I have no idea. I started going down this route and then you recommended IRSA, so I ripped all this out before doing much testing. I did notice the issue with the quotes being changed by AWS, but I never tested the functionality.
